Fixes #5888 - Limit usage of HTTP/2 connections.

sbordet opened the pull request:
* Made the high-level HttpConnectionOverHTTP2 implement MaxUsable, so it cannot be used to open more streams than allowed.
* Implemented low-level handling of explicit stream ids provided by applications.
* Implemented low-level handling of stream id overflow.
* Added test cases.

sbordet committed:
Fixes #5888 - Limit usage of HTTP/2 connections.

* Made the high-level HttpConnectionOverHTTP2 implement MaxUsable, so it cannot be used to open more streams than allowed.
* Implemented low-level handling of explicit stream ids provided by applications.
* Implemented low-level handling of stream id overflow.
* Added test cases.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban committed:
Fixes #5888 - accept existing stream ids

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
Overall, LGTM but these changes introduce a bug caught by `PriorityTest`.

sbordet committed:
Fixes #5888 - Limit usage of HTTP/2 connections.

* Now using a placeholder HTTP2Stream for streams opened with a PRIORITY, but not yet with a HEADERS.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #5888 - Limit usage of HTTP/2 connections.

* Fixed maxUsage check in AbstractConnectionPool.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-core/jetty-http2/jetty-http2-client-transport/src/main/java/org/eclipse/jetty/http2/client/transport/internal/HttpConnectionOverHTTP2.java:
```diff
@@ -108,6 +108,15 @@ public int getMaxMultiplex()
+        // clients start at 1 and increment by 2, so a
+        // connection can only be used (2^31-2)/2 times.
+        return (Integer.MAX_VALUE - 1) / 2;
+    }
```
Is there a use-case for setting a low max usage on a connection?  i.e. should this be a field, initialized to the spec number, but then we have a setter as well?

gregw commented on jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java:
```diff
@@ -836,13 +836,21 @@ protected HTTP2Stream createLocalStream(int streamId, MetaData.Request request,
         {
-            stream.setIdleTimeout(getStreamIdleTimeout());
-            flowControl.onStreamCreated(stream);
+            if (v == null || v.isPlaceHolder())
```
It is not clear how a placeholder stream can be added to the map?    Is createLocalStream called with a null request?

gregw replied:
Ah it is below.... hmmmm not exactly obvious how that happens... a placeholder is only created if newStream is null, which means there is already a non null non placeholder stream for the id.... Ah we are in a loop and you get another ID below...  a few comments might help

gregw commented on jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java:
```diff
@@ -2111,16 +2119,30 @@ private int priority(PriorityFrame frame, Callback callback)
                     frame = frame.withStreamId(streamId);
-                slot.entries = List.of(newEntry(frame, null, Callback.from(callback::succeeded, x ->
+                    // Create a placeholder stream, replaced when a follow-up HEADERS frame will be sent.
+                    stream = HTTP2Session.this.createLocalStream(streamId, null, callback::failed);
```
OK I still don't understand.  This method is creating a local stream for a specific passed in streamId, but if it discovers one already exists, then it creates another streamId and reserves a spot for a future HEADERS frame?
Why?

gregw commented on jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java:
```diff
@@ -2237,19 +2259,76 @@ private MetaData.Request extractMetaDataRequest(StreamFrame frame)
+                        // Stream id generated internally.
+                        reservedStreamId = localStreamIds.getAndAdd(2);
+                        // Check for overflow.
+                        if (reservedStreamId > 0)
```
This test is surprising to me.  I would have expected a test against a value derived from the maxUsage method, rather than just knowing that the max Id is at arithmetic overflow.

gregw reviewed (changes requested):
Sorry but the algorithm this code is trying to implement is too opaque without comments.  I can't say that it is wrong, but it looks very arbitrary, so a bit of explanation would be good.

sbordet commented:
Note to self:

* Using a placeholder stream is nice for idle timeouts, but messes with the local stream count.
* Using a side map seems simpler, but requires timeout handling

sbordet committed:
Simplified handling of streams created via priority().
Now their streams ids are stored in a side Set.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/HTTP2Test.java:
```diff
@@ -1139,6 +1140,113 @@ public void onClose(Session session, GoAwayFrame frame, Callback callback)
+            .get(5, TimeUnit.SECONDS));
+
+        // Equivalent to Integer.MAX_VALUE + 2.
+        int negativeStreamId = Integer.MIN_VALUE + 1;
```
Why not using `Integer.MAX_VALUE + 2` then?

sbordet replied:
Because it is not immediately clear that it would overflow and become negative.

sbordet committed:
Updates from review.

Introduced HTTP2Session.maxTotalLocalStreams to limit the max number of streams that might be created in a connection.

Linked this new property with the high-level AbstractConnectionPool.maxUsage.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java:
```diff
@@ -2237,19 +2265,113 @@ private MetaData.Request extractMetaDataRequest(StreamFrame frame)
-                        streamId = localStreamIds.getAndAdd(2);
-                        reserved = true;
+                        // Stream id is given.
+                        while (true)
```
This is a long maze of `if`s and `else`s in a `while` loop that partially duplicates the logic of the enclosing `if` branch, and is quite involved to follow.

Can't this be simplified and de-duplicated?

lorban reviewed (changes requested):
I think the new `reserveSlot()` logic deserves a try at simplifying it as its complexity exploded which made it rather hard to read.

Otherwise the new stream limitation LGTM.

sbordet committed:
Updates from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
